### PR TITLE
Add type annotations to pytest.fixtures.fixture

### PR DESCRIPTION
**Pull request by gtaylor**
Floating this to see if it'd be acceptable. If it looks like something that'd be good for inclusion, let me know and I can follow up with the changelog/authors commits!

### CODE REVIEW

File: src/_pytest/fixtures.py [hunk 1]
```diff
@@ -1060,14 +1066,14 @@ def _parse_fixture_args(callable_or_scope, *args, **kwargs):
...
-    callable_or_scope=None,
+    callable_or_scope: Union[Callable[..., Any], str] = None,
     *args,
-    scope="function",
-    params=None,
-    autouse=False,
-    ids=None,
-    name=None
-):
+    scope: str = "function",
```
Comment (RonnyPfannschmidt) [inline src/_pytest/fixtures.py]: just a heads up that scope is also possibly a callable on features

File: src/_pytest/fixtures.py [hunk 2]
```diff
@@ -1060,14 +1066,14 @@ def _parse_fixture_args(callable_or_scope, *args, **kwargs):
 
 
 def fixture(
-    callable_or_scope=None,
+    callable_or_scope: Union[Callable[..., Any], str] = None,
     *args,
```
Comment (asottile) [inline src/_pytest/fixtures.py]: you'll also need `*args: Any`  such that mypy will consider this function as typed

File: src/_pytest/fixtures.py [hunk 3]
```diff
@@ -1060,14 +1066,14 @@ def _parse_fixture_args(callable_or_scope, *args, **kwargs):
...
-    autouse=False,
-    ids=None,
-    name=None
-):
+    scope: str = "function",
+    params: Optional[Iterable[Any]] = None,
+    autouse: bool = False,
+    ids: Optional[List[str]] = None,
+    name: Optional[str] = None
+) -> Callable[..., Any]:
```
Comment (asottile) [inline src/_pytest/fixtures.py]: shouldn't this return `Callable[[T], T]` as it gives back a trivial decorator?

File: src/_pytest/fixtures.py [hunk 4]
```diff
@@ -1060,14 +1066,14 @@ def _parse_fixture_args(callable_or_scope, *args, **kwargs):
 
 
 def fixture(
-    callable_or_scope=None,
+    callable_or_scope: Union[Callable[..., Any], str] = None,
```
Comment (asottile) [inline src/_pytest/fixtures.py]: this should be Optional[Union[...]] as it defaults to None (`strict_optional`)

Comment (gtaylor) [reply src/_pytest/fixtures.py]: I think this is an invalid type annotation:
```
TypeError: Union[arg, ...]: each arg must be a type. Got Ellipsis.
```

Comment (asottile) [reply src/_pytest/fixtures.py]: yeah I was typing that on my phone and was hoping you'd fill in what you already had :)

File: src/_pytest/fixtures.py (hunk 3 above)
Comment (gtaylor) [reply src/_pytest/fixtures.py]: I'm still learning how to annotate things. What is `T` in this case?

Comment (asottile) [reply src/_pytest/fixtures.py]: you'll need to add `T = TypeVar('T')` -- it's a generic parameter

Comment (gtaylor) [reply src/_pytest/fixtures.py]: OK, can do. I'm a bit sketchy on the difference between this vs `Any`.

**Comment by asottile** [comment]
here's an example of the difference between a `TypeVar` and `Any`:

```python
import functools
from typing import Any
from typing import cast
from typing import Callable
from typing import TypeVar


T = TypeVar('T', bound=Callable)


def dec(func: T) -> T:
    @functools.wraps(func)
    def dec_f(*args: Any, **kwargs: Any) -> Any:
        print(f'calling {func}')
        return func(*args, **kwargs)
    return cast(T, dec_f)


def dec2(func: Any) -> Any:
    @functools.wraps(func)
    def dec_f(*args: Any, **kwargs: Any) -> Any:
        print(f'calling {func}')
        return func(*args, **kwargs)
    return dec_f


@dec
def f(x: int) -> int:
    return x ** x


@dec2
def f2(x: int) -> int:
    return x ** x


reveal_type(f)
reveal_type(f(1))

reveal_type(f2)
reveal_type(f2(1))
```

```console
$ mypy t.py
t.py:37: note: Revealed type is 'def (x: builtins.int) -> builtins.int'
t.py:38: note: Revealed type is 'builtins.int'
t.py:40: note: Revealed type is 'Any'
t.py:41: note: Revealed type is 'Any'
```

notice here that mypy preserved the signature of `f` but did not preserve the signature of `f2`

**Comment by gtaylor** [comment]
ahh, that's very helpful. Thanks for the example!

**Comment by blueyed** [comment]
Needs to be rebased.

**Comment by blueyed** [comment]
> Needs to be rebased.

I think we had some changes in this regard recently - closing for now (mainly due to feedback to my previously comment).
Please feel free to re-open / rebase etc, of course.

**Comment by blueyed** [comment]
/cc @bluetech 
Maybe we coud pick something from here?

**Pull request closed without merging**